cmd/lotus-bench/main.go

```diff
@@ -467,129 +476,162 @@ func runSeals(sb *ffiwrapper.Sealer, sbfs *basicfs.Provider, numSectors int, mid
-		}
-
-		log.Info("Generating PoRep for sector (1)")
-		c1o, err := sb.SealCommit1(context.TODO(), sid, ticket, seed.Value, pieces, cids)
-		if err != nil {
-			return nil, nil, err
-		}
-
-		sealcommit1 := time.Now()
-
-		log.Info("Generating PoRep for sector (2)")
-
-		if saveC2inp != "" {
-			c2in := Commit2In{
-				SectorNum:  int64(i),
-				Phase1Out:  c1o,
-				SectorSize: uint64(sectorSize),
-			}
-
-			b, err := json.Marshal(&c2in)
-			if err != nil {
-				return nil, nil, err
-			}
-
-			if err := ioutil.WriteFile(saveC2inp, b, 0664); err != nil {
-				log.Warnf("%+v", err)
-			}
-		}
-
-		var proof storage.Proof
-		if !skipc2 {
-			proof, err = sb.SealCommit2(context.TODO(), sid, c1o)
-			if err != nil {
-				return nil, nil, err
-			}
-		}
-
-		sealcommit2 := time.Now()
-
-		if !skipc2 {
-			svi := abi.SealVerifyInfo{
-				SectorID:              abi.SectorID{Miner: mid, Number: i},
-				SealedCID:             cids.Sealed,
-				SealProof:             sb.SealProofType(),
-				Proof:                 proof,
-				DealIDs:               nil,
-				Randomness:            ticket,
-				InteractiveRandomness: seed.Value,
-				UnsealedCID:           cids.Unsealed,
-			}
-
-			ok, err := ffiwrapper.ProofVerifier.VerifySeal(svi)
-			if err != nil {
-				return nil, nil, err
-			}
-			if !ok {
-				return nil, nil, xerrors.Errorf("porep proof for sector %d was invalid", i)
-			}
-		}
-
-		verifySeal := time.Now()
-
-		if !skipunseal {
-			log.Info("Unsealing sector")
-			{
-				p, done, err := sbfs.AcquireSector(context.TODO(), abi.SectorID{Miner: mid, Number: 1}, stores.FTUnsealed, stores.FTNone, true)
-				if err != nil {
-					return nil, nil, xerrors.Errorf("acquire unsealed sector for removing: %w", err)
-				}
-				done()
+		sealTimings[i-1].AddPiece = time.Since(start)
+	}
 
-				if err := os.Remove(p.Unsealed); err != nil {
-					return nil, nil, xerrors.Errorf("removing unsealed sector: %w", err)
+	sectorsPerWorker := numSectors / parallelism
+
+	errs := make(chan error, parallelism)
+	for wid := 0; wid < parallelism; wid++ {
+		go func(worker int) {
+			sealerr := func() error {
+				start := 1 + (worker * sectorsPerWorker)
+				end := start + sectorsPerWorker
+				for i := abi.SectorNumber(start); i < abi.SectorNumber(end); i++ {
+					ix := int(i - 1)
+					sid := abi.SectorID{
+						Miner:  mid,
+						Number: i,
+					}
+
+					start := time.Now()
+
+					trand := blake2b.Sum256(ticketPreimage)
+					ticket := abi.SealRandomness(trand[:])
+
+					log.Infof("[%d] Running replication(1)...", i)
+					pieces := []abi.PieceInfo{pieces[ix]}
+					pc1o, err := sb.SealPreCommit1(context.TODO(), sid, ticket, pieces)
+					if err != nil {
+						return xerrors.Errorf("commit: %w", err)
+					}
+
+					precommit1 := time.Now()
+
+					log.Infof("[%d] Running replication(2)...", i)
+					cids, err := sb.SealPreCommit2(context.TODO(), sid, pc1o)
+					if err != nil {
+						return xerrors.Errorf("commit: %w", err)
+					}
```

Comment:
Running precommit2/commit2 in parallel may not yield the best perf, we should probably have a way to limit how many precommit2s can be run in parallel (they can be run in parallel with pc1, since pc2 is now mostly GPU)